src/utils/FileBuffer.cpp

```diff
@@ -0,0 +1,334 @@
+  const ExpectVoid resB = _saveRemainder();
+  if (!resB.has_value()) {
+    return ft::unexpected<BufferException>(resB.error().what());
+  }
+
+  const ExpectVoid resC = _calcSize();
+  if (!resC.has_value()) {
+    return ft::unexpected<BufferException>(resC.error().what());
+  }
+
+  return front;
+}
+
+std::size_t FileBuffer::size() const
+{
+  return _size;
+}
+
+// NOLINTBEGIN(bugprone-random-generator-seed, misc-predictable-rand)
+static std::string getRandomeFileName()
+{
+  static bool seeded = false;
+
+  if (!seeded) {
+    seeded = true;
+    std::srand(std::time(FT_NULLPTR));
+  }
+
+  std::string filename;
+  filename.append("tmpfile_");
+  filename.append(ft::to_string(std::rand()));
+  filename.append(ft::to_string(std::rand()));
+  return filename;
+}
+// NOLINTEND(bugprone-random-generator-seed, misc-predictable-rand)
+
+IBuffer::ExpectVoid FileBuffer::openTmpFile()
+{
+  assert(!_fs.is_open());
+  const std::size_t countMax = 100;
+  std::size_t count = 0;
+  while (count < countMax) {
+    _fileName = getRandomeFileName();
+    _fs.open(_fileName.c_str(), std::ios::in);
+    // check if file does not exist
+    if (!_fs.is_open()) {
+      // create file
+      _fs.open(_fileName.c_str(),
+               std::ios::in | std::ios::out | std::ios::binary |
+                 std::ios::trunc);
+      if (_fs.is_open()) {
+        _size = 0;
+        return ExpectVoid();
+      }
+    } else {
+      _fs.close();
+    }
+    ++count;
+  }
+  // tried to open 100 random files
+  return ft::unexpected<BufferException>(errOpen);
+}
+
+/* ************************************************************************** */
+// PRIVATE
+
+IBuffer::ExpectStr FileBuffer::_getFront(std::size_t bytes)
+{
+  if (!_fs.is_open()) {
+    return ft::unexpected<BufferException>(errFileNotOpened);
+  }
+
+  std::string front(bytes, '\0');
+  _fs.read(&front[0], static_cast<std::streamsize>(bytes));
+  const std::streamsize actuallyRead = _fs.gcount();
+  if (actuallyRead == 0) {
+    return ft::unexpected<BufferException>(errFileEmpty);
+  }
+  front.resize(static_cast<std::size_t>(actuallyRead));
+  return front;
+}
+
+IBuffer::ExpectVoid FileBuffer::_saveRemainder()
+{
+  // Create a temporary file for the remainder
+  FileBuffer tmp;
+  ExpectVoid res = tmp.openTmpFile();
+  if (!res.has_value()) {
+    return res;
+  }
+
+  // Copy the remaining data
+  res = _copyData(_fs, tmp._fs);
+  if (!res.has_value()) {
+    return res;
+  }
+
+  res = _replaceCurrFile(tmp);
+  if (!res.has_value()) {
+    return res;
+  }
+  return ExpectVoid();
+}
+
+// NOLINTBEGIN(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+IBuffer::ExpectVoid FileBuffer::_copyData(std::fstream& bufFrom,
+                                          std::fstream& bufTo)
+{
+  char buffer[_copyBufferSize];
+  while (bufFrom.good()) {
+    bufFrom.read(buffer, _copyBufferSize);
+    // if (bufFrom.fail()) {
+    //   return ft::unexpected<BufferException>(errRead);
+    // }
```

Comment:
This should work:
```suggestion
  while (!bufFrom.eof()) {
    bufFrom.read(buffer, _copyBufferSize);
    if (bufFrom.fail() && !bufFrom.eof()) {
      return ft::unexpected<BufferException>(errRead);
    }
```

> end of file condition occurs on the input sequence (**in which case, `setstate(failbit|eofbit)` is called**). The number of successfully extracted characters can be queried using gcount().

https://en.cppreference.com/w/cpp/io/basic_istream/read.html